### PR TITLE
fix(crdt): YText.Format formatText port — stop stripping formatting outside the range (#123)

### LINKED ISSUE #123: crdt: YText.Format over-deletes the closing format marker when the range ends inside an un-split run
## Summary

`YText.Format(txn, index, length, attrs)` deletes the closing format marker that bounds content **after** the formatted range when the range end falls inside an un-split text run. Formatting on content outside `[index, index+length)` is destroyed.

## Mechanism

The pre-insert cleanup loop in `Format` ([crdt/ytext.go:718-746](https://github.com/reearth/ygo/blob/main/crdt/ytext.go)) deletes every same-key `ContentFormat` it walks past, but the `consumed >= length` break guard is only in the text/embed branch — not the `ContentFormat` branch. When the range start split is a no-op (offset 0) and the whole run is consumed in one step, `consumed` jumps past `length`, and the loop then reaches and deletes the **closing** marker at `index+length`, which belongs to content outside the range.

## Reproduction

Confirmed failing on current `main`:

- Insert `"aaabbb"`, bold `[0,6)`, then bold `[0,3)` → expected all bold, actual `{"aaa" bold}{"bbb" plain}`.
- Insert `"abc DEF ghi"`, bold `[4,7)` (`DEF`), then bold `[0,4)` → `DEF` loses its bold.
- Insert `"aaabbb"`, bold `[0,6)`, then unbold `[2,4)` → chars 4-5 lose bold.

Cases where the range end coincides with an already-split run boundary pass — the bug only fires when the range end is interior to a single un-split run with a same-key marker beyond it.

## Fix direction

Make the cleanup loop respect the range end for `ContentFormat` items too: stop deleting markers once `consumed >= length` (delete only markers strictly within the range), and account for the un-split-run case so the trailing closer at `index+length` is preserved. Add ToDelta regression tests for the boundary scenarios above.

### PR DESCRIPTION
**Pull request by bnimit**
## Summary

Fixes **#123**: `YText.Format` stripped formatting from content *outside* the
range it was applied to. Re-applying or toggling a format over a sub-range of an
already-formatted run deleted the closing marker that bounded content after
`index+length`, so the surrounding run lost its formatting. This was a verified
bug on current `main` (fresh local docs), and — per **[yjs#606](https://github.com/yjs/yjs/issues/606)** — bites hardest on documents loaded via `ApplyUpdate` (un-split runs from the wire).

## The fix

`YText.Format` is now a faithful port of the Yjs JS `formatText` algorithm
(validated against `yjs@13.6.30`, the version the compat suite runs):

- A cursor (`itemTextPos`, ≡ Yjs `ItemTextListPosition`) walks the range, tracking the format state.
- `minimizeAttributeChanges` skips leading markers that already match the target.
- `insertAttributes` opens a marker only where the value actually changes, recording the **negated** (restore-after-range) value.
- The loop deletes only *in-range* overlapping markers and continues past `length==0` over trailing markers to drop redundant ones — this bounds marker accumulation (the previous code's over-aggressive in-range deletion *was* the accumulation guard, which is what caused the over-delete).
- `insertNegatedAttributes` restores the post-range state.

`YText.ToDelta` now **coalesces adjacent inserts with equal attributes** into one
op (matching Yjs `toDelta`), required so a run split across Items by a `Format`
boundary renders an identical delta.

## Why this approach (validation)

I reverted a first, simpler attempt that fixed 3/5 cases but broke overlap-at-boundary and reintroduced marker accumulation — the correct fix needs the full negated-attributes algorithm, not a one-line guard change. I then cross-checked the canonical Yjs source and confirmed empirically that `yjs@13.6.30` produces the expected deltas on both fresh and `applyUpdate`-loaded docs (#606 is fixed there), so matching it = correctness.

## Testing

- 7 Go scenario tests: rebold sub-range, overlap-at-boundary, unset-middle, unset-prefix, already-in-effect no-op, and a **loaded-via-`ApplyUpdate`** case (the #606 path).
- New **Go→JS compat test** (`TestCompat_FormatNegatedAttrs_GoToJS`): real Yjs decodes ygo's encoded document and asserts the `toDelta` matches across all scenarios.
- Existing `RepeatedToggles_DoNotInflateStore` accumulation bound stays green; one existing `ToDelta` sanity assertion updated to the (now Yjs-faithful) coalesced shape.
- `go test -race ./...`, `go vet`, `gofmt`, `golangci-lint` all clean.

## Compatibility

No breaking API changes. `ToDelta` returns the same text + attributes, just
merged into fewer ops where they were previously split per backing Item — code
that compares the full op list (rather than rendered content) should expect the
coalesced shape. Ships as **v1.27.0**.

Closes #123.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crdt/ytext.go [hunk 1]
```diff
@@ -697,139 +697,216 @@ func (txt *YText) Format(txn *Transaction, index, length int, attrs Attributes)
...
+		if pos.right.Deleted {
+			// advance
+		} else if cf, ok := pos.right.Content.(*ContentFormat); ok && attrEqual(attrs[cf.Key], cf.Val) {
+			// redundant marker already matching the target — advance
 		} else {
-			endOriginRight = nil
+			break
 		}
+		pos.forward()
 	}
```
[the remark this answers is not in the document]
Comment (bnimit) [reply crdt/ytext.go:866]: This matches Yjs `minimizeAttributeChanges` exactly: `equalAttrs(attributes[key] ?? null, value)` also skips null-valued markers for keys absent from `attributes`. Advancing the cursor past them is intentional in the reference (new markers integrate after them), and the Go→JS compat test (`TestCompat_FormatNegatedAttrs_GoToJS`) confirms parity against yjs@13.6.30. Keeping as-is to avoid diverging from Yjs.

File: crdt/ytext.go [hunk 2]
```diff
@@ -905,42 +982,58 @@ func (txt *YText) ToDelta() []Delta {
...
+			flush()
+			deltas = append(deltas, Delta{Op: DeltaOpInsert, Insert: c.Val, Attributes: attrsCopy()})
 		case *ContentFormat:
-			if c.Val == nil {
-				delete(currentAttrs, c.Key)
-			} else {
-				currentAttrs[c.Key] = c.Val
-			}
+			updateAttr(currentAttrs, c)
 		}
```
[the remark this answers is not in the document]
Comment (bnimit) [reply crdt/ytext.go:1054]: Done (9e69c32). `ToDelta` now tracks a dirty flag set only when a live `ContentFormat` changes `currentAttrs`, so a contiguous run with no intervening markers does zero per-item map copies/compares. I kept the compare-by-effective-attributes coalescing (rather than flush-on-every-`ContentFormat`) so a redundant marker doesn't over-split a run relative to Yjs.
